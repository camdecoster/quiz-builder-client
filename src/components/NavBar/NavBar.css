nav {
    width: 100%;
    margin-bottom: 10px;
    padding-bottom: 10px;
    /* position: sticky; */
    position: fixed;
    top: 0;
    left: 0;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    border-bottom: 2px solid var(--border);
    border-bottom-left-radius: 15px;
    border-bottom-right-radius: 15px;
    background-color: var(--background);
    z-index: 500;
}

.sticky {
    /* transform: translateY(-160px);
    transition: transform 0.5s; */
    animation: moveDown 0.5s ease-in-out;
}

.not_sticky {
    /* transform: translateY(-160px); */
    display: none;
    /* animation: moveUp 0.5s ease-in-out; */
}

nav .container_flex {
    width: 90%;
    min-width: 300px;
    display: flex;
    flex-direction: row;
    justify-content: space-around;
    align-items: center;
}

nav h1 {
    font-size: 20px;
    margin-top: 8px;
    margin-bottom: 8px;
}

nav a {
    text-decoration: none;
}

nav p {
    display: inline;
}

nav .link_button {
    height: 40px;
    width: 40px;
    border-radius: 20px;
    margin-left: 10px;
    margin-right: 10px;
}

nav .link_button .faIcon {
    font-size: 20px;
}

nav .link_list {
    display: flex;
    flex-direction: row;
}

#logged_in_note {
    margin-top: 5px;
    margin-bottom: 0;
    font-size: 12px;
}

@keyframes moveDown {
    from {
        transform: translateY(-160px);
    }
    to {
        transform: translateY(0);
    }
}

@keyframes moveUp {
    from {
        transform: translateY(0);
    }
    to {
        transform: translateY(-160px);
    }
}

@media screen and (min-width: 400px) {
    nav h1 {
        font-size: 25px;
    }

    nav .link_button {
        height: 80px;
        width: 80px;
        border-radius: 15px;
        margin-left: 15px;
        margin-right: 15px;
    }

    nav .link_button .faIcon {
        font-size: 28px;
        margin-bottom: 8px;
    }

    nav .button_label {
        font-size: 12px;
    }
}
