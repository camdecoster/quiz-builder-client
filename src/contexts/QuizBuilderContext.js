import React from "react";

const QuizBuilderContext = React.createContext({
    dateCurrent: null,
    quizzes: null,
    setClassNames: () => {},
    setDoFetchData: () => {},
    setQuizzes: () => {},
    toggleClassNames: () => {},
});

export default QuizBuilderContext;
