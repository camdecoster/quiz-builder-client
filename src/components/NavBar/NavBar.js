// React
import React, { useContext } from "react";
import { Link, NavLink, useHistory } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

// Configuration
import "./NavBar.css";
import QuizBuilderContext from "../../contexts/QuizBuilderContext";
import TokenService from "../../services/token-service";

// Components
import AddItemLinkButton from "../Utilities/AddItemLinkButton/AddItemLinkButton";

export default function NavBar() {
    // Access context
    const context = useContext(QuizBuilderContext);

    // Access history
    const history = useHistory();

    // Handle tasks when logging out
    function handleLogOutClick() {
        // Clear auth token
        TokenService.clearAuthToken();

        // Route to homepage
        history.push("/");

        // Clear out User information
        // context.setExpenses([]);
        // context.setPayment_methods([]);
        // context.setCategories([]);
    }

    function createNavLinkList(navLinkTargets) {
        // Build list of NavLinks for <ul> in SideBar
        let navLinks = "";

        navLinks = navLinkTargets.map((navLink) => {
            return (
                <AddItemLinkButton
                    to={"/" + navLink.link}
                    name={navLink.name}
                    icon={navLink.icon}
                    key={navLink.name}
                />
            );
        });

        return navLinks;
    }

    function renderAuthorized() {
        const navLinkTargets = [
            // {
            //     name: "Dashboard",
            //     link: "",
            // },
            {
                name: "New Quiz",
                link: "quizzes/edit/new",
                icon: "plus-circle",
            },
            {
                name: "Quizzes",
                link: "quizzes",
                icon: "list-alt",
            },
        ];

        // Get NavLinks for NavBar
        const navLinks = createNavLinkList(navLinkTargets);

        // Add log out button
        navLinks.push(
            <button
                className='link_button'
                type='button'
                title='Log Out'
                key='Log Out'
                aria-label='Log Out'
                onClick={() => handleLogOutClick()}
            >
                <FontAwesomeIcon className='faIcon' icon='sign-out-alt' />
            </button>
        );

        return navLinks;
    }

    function renderUnauthorized() {
        const navLinkTargets = [
            {
                name: "Log In",
                link: "login",
                icon: "sign-in-alt",
            },
            {
                name: "Register",
                link: "register",
                icon: "plus-square",
            },
        ];

        return createNavLinkList(navLinkTargets);
    }

    return (
        <nav role='navigation'>
            <h1>
                <Link
                    to='/'
                    title={
                        TokenService.hasAuthToken()
                            ? "Go to Dashboard"
                            : "Go to Main Page"
                    }
                >
                    Quiz Builder
                </Link>
            </h1>
            <div className='link_list'>
                {/* Links and logout should only be shown if user logged in */}
                {TokenService.hasAuthToken()
                    ? renderAuthorized()
                    : renderUnauthorized()}
            </div>
            {!!TokenService.getUser() ? (
                <p id='logged_in_note'>
                    {`Logged in as ${TokenService.getUser()}`}
                </p>
            ) : (
                ""
            )}
        </nav>
    );
}
