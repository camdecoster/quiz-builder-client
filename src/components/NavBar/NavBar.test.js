// React
import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";

// Configuration
import QuizBuilderContext from "../../contexts/QuizBuilderContext";
import TokenService from "../../services/token-service";

// Components
import NavBar from "./NavBar";

// Create mock service for TokenService
jest.mock("../../services/token-service");

function NavBarTest(props) {
    return (
        <MemoryRouter>
            <QuizBuilderContext.Provider
                value={{ showUserMenu: props.showUserMenu }}
            >
                <NavBar />
            </QuizBuilderContext.Provider>
        </MemoryRouter>
    );
}

describe("<NavBar> component", () => {
    it("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(<NavBarTest showUserMenu={false} />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    describe("renders the UI", () => {
        describe("with authorization", () => {
            TokenService.hasAuthToken.mockReturnValue(true);

            it("with the UserMenu showing", () => {
                const tree = renderer
                    .create(<NavBarTest showUserMenu={true} />)
                    .toJSON();
                expect(tree).toMatchSnapshot();
            });

            it("with the UserMenu showing", () => {
                const tree = renderer
                    .create(<NavBarTest showUserMenu={false} />)
                    .toJSON();
                expect(tree).toMatchSnapshot();
            });
        });

        describe("without authorization", () => {
            TokenService.hasAuthToken.mockReturnValue(false);

            it("with the UserMenu showing", () => {
                const tree = renderer
                    .create(<NavBarTest showUserMenu={true} />)
                    .toJSON();
                expect(tree).toMatchSnapshot();
            });

            it("with the UserMenu showing", () => {
                const tree = renderer
                    .create(<NavBarTest showUserMenu={false} />)
                    .toJSON();
                expect(tree).toMatchSnapshot();
            });
        });
    });
});
