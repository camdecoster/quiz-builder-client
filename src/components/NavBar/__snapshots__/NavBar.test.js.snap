// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavBar> component renders the UI with authorization with the UserMenu showing 1`] = `
<nav
  className=""
  role="navigation"
>
  <h1>
    <a
      href="/"
      onClick={[Function]}
      title="Go to Main Page"
    >
      QUIZ BUILDER
    </a>
  </h1>
  <div
    className="link_list"
  >
    <div
      id="AddItemLinkButton"
    >
      <a
        aria-label="Go to Login Page"
        href="/login"
        onClick={[Function]}
        title="Go to Login Page"
      >
        <button
          aria-label="Go to Login Page"
          className="link_button "
          type="button"
        >
          <div
            className="button_label hide_label"
            style={null}
          >
            LOG IN
          </div>
        </button>
      </a>
    </div>
    <div
      id="AddItemLinkButton"
    >
      <a
        aria-label="Go to Register Page"
        href="/register"
        onClick={[Function]}
        title="Go to Register Page"
      >
        <button
          aria-label="Go to Register Page"
          className="link_button "
          type="button"
        >
          <div
            className="button_label hide_label"
            style={null}
          >
            REGISTER
          </div>
        </button>
      </a>
    </div>
    <div
      id="AddItemLinkButton"
    >
      <button
        aria-label="Try a random quiz"
        className="link_button "
        onClick={[Function]}
        type="button"
      >
        <div
          className="button_label hide_label"
          style={null}
        >
          RANDOM
        </div>
      </button>
    </div>
  </div>
  
</nav>
`;

exports[`<NavBar> component renders the UI with authorization with the UserMenu showing 2`] = `
<nav
  className=""
  role="navigation"
>
  <h1>
    <a
      href="/"
      onClick={[Function]}
      title="Go to Main Page"
    >
      QUIZ BUILDER
    </a>
  </h1>
  <div
    className="link_list"
  >
    <div
      id="AddItemLinkButton"
    >
      <a
        aria-label="Go to Login Page"
        href="/login"
        onClick={[Function]}
        title="Go to Login Page"
      >
        <button
          aria-label="Go to Login Page"
          className="link_button "
          type="button"
        >
          <div
            className="button_label hide_label"
            style={null}
          >
            LOG IN
          </div>
        </button>
      </a>
    </div>
    <div
      id="AddItemLinkButton"
    >
      <a
        aria-label="Go to Register Page"
        href="/register"
        onClick={[Function]}
        title="Go to Register Page"
      >
        <button
          aria-label="Go to Register Page"
          className="link_button "
          type="button"
        >
          <div
            className="button_label hide_label"
            style={null}
          >
            REGISTER
          </div>
        </button>
      </a>
    </div>
    <div
      id="AddItemLinkButton"
    >
      <button
        aria-label="Try a random quiz"
        className="link_button "
        onClick={[Function]}
        type="button"
      >
        <div
          className="button_label hide_label"
          style={null}
        >
          RANDOM
        </div>
      </button>
    </div>
  </div>
  
</nav>
`;

exports[`<NavBar> component renders the UI without authorization with the UserMenu showing 1`] = `
<nav
  className=""
  role="navigation"
>
  <h1>
    <a
      href="/"
      onClick={[Function]}
      title="Go to Main Page"
    >
      QUIZ BUILDER
    </a>
  </h1>
  <div
    className="link_list"
  >
    <div
      id="AddItemLinkButton"
    >
      <a
        aria-label="Go to Login Page"
        href="/login"
        onClick={[Function]}
        title="Go to Login Page"
      >
        <button
          aria-label="Go to Login Page"
          className="link_button "
          type="button"
        >
          <div
            className="button_label hide_label"
            style={null}
          >
            LOG IN
          </div>
        </button>
      </a>
    </div>
    <div
      id="AddItemLinkButton"
    >
      <a
        aria-label="Go to Register Page"
        href="/register"
        onClick={[Function]}
        title="Go to Register Page"
      >
        <button
          aria-label="Go to Register Page"
          className="link_button "
          type="button"
        >
          <div
            className="button_label hide_label"
            style={null}
          >
            REGISTER
          </div>
        </button>
      </a>
    </div>
    <div
      id="AddItemLinkButton"
    >
      <button
        aria-label="Try a random quiz"
        className="link_button "
        onClick={[Function]}
        type="button"
      >
        <div
          className="button_label hide_label"
          style={null}
        >
          RANDOM
        </div>
      </button>
    </div>
  </div>
  
</nav>
`;

exports[`<NavBar> component renders the UI without authorization with the UserMenu showing 2`] = `
<nav
  className=""
  role="navigation"
>
  <h1>
    <a
      href="/"
      onClick={[Function]}
      title="Go to Main Page"
    >
      QUIZ BUILDER
    </a>
  </h1>
  <div
    className="link_list"
  >
    <div
      id="AddItemLinkButton"
    >
      <a
        aria-label="Go to Login Page"
        href="/login"
        onClick={[Function]}
        title="Go to Login Page"
      >
        <button
          aria-label="Go to Login Page"
          className="link_button "
          type="button"
        >
          <div
            className="button_label hide_label"
            style={null}
          >
            LOG IN
          </div>
        </button>
      </a>
    </div>
    <div
      id="AddItemLinkButton"
    >
      <a
        aria-label="Go to Register Page"
        href="/register"
        onClick={[Function]}
        title="Go to Register Page"
      >
        <button
          aria-label="Go to Register Page"
          className="link_button "
          type="button"
        >
          <div
            className="button_label hide_label"
            style={null}
          >
            REGISTER
          </div>
        </button>
      </a>
    </div>
    <div
      id="AddItemLinkButton"
    >
      <button
        aria-label="Try a random quiz"
        className="link_button "
        onClick={[Function]}
        type="button"
      >
        <div
          className="button_label hide_label"
          style={null}
        >
          RANDOM
        </div>
      </button>
    </div>
  </div>
  
</nav>
`;
